scipy/optimize/_quasi_newton_approx.py

```diff
@@ -0,0 +1,261 @@
+            in two points: ``delta_grad = grad(x2) - grad(x1)``.
+        """
+        # Auxiliar variables w and z
+        if self.approx_type == 'hess':
+            w = delta_x
+            z = delta_grad
+        else:
+            w = delta_grad
+            z = delta_x
+        # Do some common operations
+        wz = np.dot(w, z)
+        Mw = self.dot(w)
+        wMw = Mw.dot(w)
+        if wMw == 0.0:
+            return
+        # Check if curvature condition is violated
+        if wz < self.min_curvature * wMw:
+            # If the option 'skip_update' is set
+            # we just skip the update when the condion
+            # is violated.
+            if self.exception_strategy == 'skip_update':
+                return
+            # If the option 'damped_bfgs' is set we
+            # interpolate between the actual BFGS
+            # result and the unmodified matrix.
+            elif self.exception_strategy == 'damped_bfgs':
+                update_factor = (1-self.min_curvature) / (1 - wz/wMw)
+                z = update_factor*z + (1-update_factor)*Mw
+                wz = np.dot(w, z)
+                Mw = self.dot(w)
+                wMw = Mw.dot(w)
+        # Update matrix
+        if self.approx_type == 'hess':
+            return self._update_hessian(wz, Mw, wMw, z)
+        else:
+            return self._update_inverse_hessian(wz, Mw, wMw, z)
+
+
+class SR1(QuasiNewtonApprox):
+    """Symmetric-rank-1 Hessian matrix approximation.
+
+    Parameters
+    ----------
+    min_denominator : float
+        Define the minimum allowed value of the denominator
+        in the update. When the condition is violated we skip
+        the update. By default uses ``1e-8``.
+
+    References
+    ----------
+    .. [1] Nocedal, Jorge, and Stephen J. Wright. "Numerical optimization"
+           Second Edition (2006).
+    """
+
+    def __init__(self, min_denominator=1e-8):
+        self.min_denominator = min_denominator
+        super(SR1, self).__init__()
+
+    def update(self, delta_x, delta_grad):
+        """Update approximation matrix.
+
+        Parameters
+        ----------
+        delta_x : ndarray
+            The difference between two points the gradient
+            function have been evaluated: ``delta_x = x2 - x1``.
+        delta_grad : ndarray
+            The difference between the gradient evaluated
+            in two points: ``delta_grad = grad(x2) - grad(x1)``.
+        """
+        # Auxiliar variables w and z
+        if self.approx_type == 'hess':
+            w = delta_x
+            z = delta_grad
+        else:
+            w = delta_grad
+            z = delta_x
+        # Do some common operations
+        Mw = self.dot(w)
+        z_minus_Mw = z - Mw
+        denominator = np.dot(w, z_minus_Mw)
+        # If the denominator is too small
+        # we just skip the update.
+        if np.abs(denominator) < self.min_denominator*norm(w)*norm(z_minus_Mw) \
+           or denominator == 0.0:
```

Comment:
Check `denominator == 0.0` looks unnecessary.

Reply:
The way it is now, it is actually necessary since the case for which ``denominator == 0.0`` is not entirely contemplated by the previous inequality. I had some trouble with this when testing the method.

Nevertheless, now that you have mentioned it, it is an ugly expression.  I will replace it with:
```
if np.abs(denominator) <= self.min_denominator*norm(w)*norm(z_minus_Mw)
```
which will have basically the same effect.

Reply:
Actually that is what I meant, that `np.abs(....)` check is sufficient.